src/lib/descriptor/mod.rs

```diff
@@ -14,69 +14,175 @@ mod cargo_alias;
+    let mut graph: GraphMap<&str, (), _> = DiGraphMap::new();
+
+    let keys: HashSet<_> = combined.iter().map(|(key, _)| *key).collect();
+    for key in keys {
+        graph.add_node(key);
     }
 
-    for (key, value) in extended.iter() {
-        let key_str = key.to_string();
+    debug!("initial graph: {graph:?}");
 
-        if !key_str.starts_with("CARGO_MAKE_CURRENT_TASK_") {
-            let value_clone = value.clone();
+    for (key, val) in &combined {
+        // combined is unique (only latest value),
+        // which is why we do not need to delete any previously declared outbound edges.
 
-            match merged.get(&key_str) {
-                Some(base_value) => {
-                    let base_value_clone = base_value.clone();
-                    match (base_value_clone, value_clone.clone()) {
-                        (
-                            EnvValue::Profile(ref base_profile_env),
-                            EnvValue::Profile(ref extended_profile_env),
-                        ) => {
-                            let mut base_profile_env_mut = base_profile_env.clone();
-                            let mut extended_profile_env_mut = extended_profile_env.clone();
+        // if the env variable is in the current scope add add an edge,
+        // otherwise it is referencing and external variable.
+        for used in merge_env_depends_on(val).into_iter() {
+            if graph.contains_node(used) {
+                graph.add_edge(*key, used, ());
+            }
+        }
+    }
 
-                            let merged_sub_env =
-                                merge_env(&mut base_profile_env_mut, &mut extended_profile_env_mut);
+    let variables = match toposort(&graph, None) {
+        Ok(iter) => iter,
+        Err(_) => {
+            // cycle has been detected, for better performance we now only
+            // execute scc.
+            // In strongly-connected-components every vertex
+            // (node) is reachable from every other node.
+            // This means that there **must** be a cycle.
+            // This isn't strictly necessary, but aids when debugging.
+            let mut err = "A cycle between different env variables has been detected.".to_owned();
+            for scc in kosaraju_scc(&graph) {
+                let render = scc
+                    .iter()
+                    .chain(scc.first())
+                    .map(ToString::to_string)
+                    .reduce(|acc, name| format!("{acc} -> {name}"));
+
+                if let Some(render) = render {
+                    err.push_str(&format!(" Cycle: {}.", render));
+                }
+            }
 
-                            merged.insert(key_str, EnvValue::Profile(merged_sub_env));
-                        }
-                        _ => {
-                            merged.insert(key_str, value_clone);
-                        }
-                    };
+            return Err(err);
```

Comment:
breaking the build due to a cycle is a new behavior. I hope this won't break other projects.

Reply:
I don't think so. I believe, in general, this is an improvement because it tells the user specifically why a configuration won't work. Pre PR, this would just silently error out, meaning it would lead to more frustration if that makes sense.

Reply:
agree with approach. just concerned. later, i'm getting lots of issues from unhappy people :)

Reply:
I see. This is why specifically search for strongly connected components so that people can see **where** a cycle is.

So if you declare

```
ENV2 = ${ENV1}
ENV1 = ${ENV2}
```

then it will tell you: cycle detected: ENV2 -> ENV1 -> ENV2.

I think this should help minimize the number of issues. Maybe I can clarify the issue message even more, to redirect to the README for a link explaining how this works.

Reply:
cool